ext/CEGUI/IrrlichtBaWTexture.cpp

```diff
@@ -220,72 +152,117 @@ void IrrlichtBaWTexture::loadFromMemory(const void* buffer, const CEGUI::Sizef&
     region.imageSubresource.layerCount = 1u;
     region.bufferOffset = 0u;
     region.bufferRowLength = 0;
-    region.bufferImageHeight = 0u; 
+    region.bufferImageHeight = 0u;
     region.imageOffset = { 0u, 0u, 0u };
     region.imageExtent = imageParams.extent;
 
-    auto gpuImage = device->getVideoDriver()->createGPUImage(std::move(imageParams));
-    driver->copyBufferToImage(gpuPixelBuffer.get(), gpuImage.get(), regions->size(), regions->begin());
+    driver->copyBufferToImage(gpuBuffer.get(), gpuImage.get(), regions->size(), regions->begin());
 
     asset::IImageView<video::IGPUImage>::SCreationParams viewParams;
     viewParams.flags = static_cast<decltype(viewParams.flags)>(0);
     viewParams.format = newFormat;
     viewParams.image = gpuImage;
     viewParams.subresourceRange.baseArrayLayer = 0;
     viewParams.subresourceRange.baseMipLevel = 0;
-    viewParams.subresourceRange.layerCount = 1;
-    viewParams.subresourceRange.levelCount = 1;
+    viewParams.subresourceRange.layerCount = imageParams.arrayLayers;
+    viewParams.subresourceRange.levelCount = imageParams.mipLevels;
     viewParams.viewType = decltype(viewParams.viewType)::ET_2D;
 
-    frameBuffer = createFrameBuffer(driver->createGPUImageView(std::move(viewParams)));
+    gpuImageViewTexture = driver->createGPUImageView(std::move(viewParams));
 }
 
 void IrrlichtBaWTexture::blitFromMemory(const void* sourceData, const CEGUI::Rectf& area)
 {
-    bool status = frameBuffer;
+    bool status = gpuImageViewTexture.get();
 
     if (status)
     {
-        auto gpuTexture = frameBuffer->getAttachment(textureColorAttachment);
-        auto& viewParams = gpuTexture->getCreationParameters();
+        auto& viewParams = gpuImageViewTexture->getCreationParameters();
         auto& extent = viewParams.image->getCreationParameters().extent;
-        irr::core::recti coverArea(area.left(), area.top(), area.getWidth(), area.getHeight());
 
-        auto temporaryInFrameBuffer = createFrameBuffer(sourceData, viewParams.format, extent.width, extent.height);
-        driver->blitRenderTargets(temporaryInFrameBuffer, frameBuffer, false, false, coverArea, coverArea);
-        temporaryInFrameBuffer->drop();
+        auto requirements = getMemoryRequirements(area.getWidth() * area.getHeight() * asset::getTexelOrBlockBytesize(viewParams.format));
+
+        auto gpuBuffer = driver->createGPUBufferOnDedMem(requirements);
+        driver->updateBufferRangeViaStagingBuffer(gpuBuffer.get(), 0, requirements.vulkanReqs.size, sourceData);
```

Comment:
you can `copyBufferToImage` straight from the streaming/staging buffer... here you're creating one buffer too many